zeus21/correlations.py

```diff
@@ -298,225 +445,458 @@ def __init__(self, Cosmo_Parameters, ClassCosmo, Correlations, T21_coefficients,
+        for ir in range(len(T21_coefficients.Rtabsmoo)):
+            corrdNL = corrdNLBIG[:,:,:,:,ir]
+            ###NUMEXPR Tests
+            gammaTimesCorrdNL = ne.evaluate('gammamatrixR1R2 * corrdNL')
+            expGammaCorrMinusLinear = ne.evaluate('exp(gammaTimesCorrdNL) - 1 - gammaTimesCorrdNL')
+
+            deltaXiTxAddend = ne.evaluate('coeffsTxALL * expGammaCorrMinusLinear')
+            deltaXiTxAddend = np.einsum('ijkl->ik', deltaXiTxAddend, optimize = True) #equivalent to np.sum(deltaXiTxAddend, axis = (1, 3))
+            deltaXiTxAddend = np.cumsum(deltaXiTxAddend[::-1], axis = 0)[::-1]
+            deltaXiTxAddend = np.moveaxis(deltaXiTxAddend, 1, 0)
+            deltaXiTxAddend = np.cumsum(deltaXiTxAddend[::-1], axis = 0)[::-1]
+            self._II_deltaxi_Tx[:,ir] = np.einsum('ii->i', deltaXiTxAddend, optimize = True)
+
+            deltaXiXaTxAddend = ne.evaluate('coeffsXaTxALL * expGammaCorrMinusLinear')
+            deltaXiXaTxAddend = np.einsum('ijkl->ik', deltaXiXaTxAddend, optimize = True) #equivalent to np.sum(deltaXiXaTxAddend, axis = (1, 3))
+            deltaXiXaTxAddend = np.moveaxis(deltaXiXaTxAddend, 1, 0)
+            deltaXiXaTxAddend = np.cumsum(deltaXiXaTxAddend[::-1], axis = 0)[::-1]
+            self._II_deltaxi_xaTx[:,ir] = np.einsum('ii->i', deltaXiXaTxAddend, optimize = True)
+
+
+        self._II_deltaxi_Tx *= np.array([_coeffTx_units]).T**2
+        self._II_deltaxi_xaTx *= np.array([coeffzp1xa * _coeffTx_units]).T
+
+
+        if (constants.FLAG_DO_DENS_NL):
+            D_coeffR2Tx = coeffR2Tx.reshape(1, *coeffR2Tx.shape, 1)
+            D_coeffzp2Tx = coeffzp2Tx.flatten().reshape(1, *coeffzp2Tx.flatten().shape, 1)
+            D_gammaR2 = gammaR2.reshape(1, *gammaR2.shape , 1)
+            D_growthRmatrix = growthRmatrix[:,0].reshape(*growthRmatrix[:,0].shape, 1, 1, 1)
+            D_corrdNL = corrdNLBIG.squeeze()[0].reshape(1, 1, *corrdNLBIG.squeeze()[0].shape)
+
+            self._II_deltaxi_dTx =  D_coeffzp2Tx * np.sum(D_coeffR2Tx * ((np.exp(D_gammaR2 * D_growthRmatrix * D_corrdNL)-1.0) - D_gammaR2 * D_growthRmatrix * D_corrdNL), axis = 2)
+
+            self._II_deltaxi_dTx = np.moveaxis(self._II_deltaxi_dTx, 1, 0)
+            self._II_deltaxi_dTx = np.cumsum(self._II_deltaxi_dTx[::-1], axis = 0)[::-1]
+            self._II_deltaxi_dTx = np.moveaxis(self._II_deltaxi_dTx, 1, 0)
+            self._II_deltaxi_dTx = np.einsum('iik->ik', self._II_deltaxi_dTx, optimize = True)
+            self._II_deltaxi_dTx *= np.array([_coeffTx_units]).T
+            
+        return 1
+
+    def get_all_corrs_IIxIII(self, Cosmo_Parameters, Correlations, T21_coefficients):
+        "Returns the correlation function of all observable at each z in zintegral"
+        #HectorAfonsoCruz: I deleted the bubbles and EoR part, to be done later.....
         #_iRnonlinear = np.arange(Cosmo_Parameters.indexminNL,Cosmo_Parameters.indexmaxNL)
+
+        zGreaterMatrix100 = np.copy(T21_coefficients.zGreaterMatrix)
```

Comment:
I'm not sure we can do that with enough repeated lines to sufficiently decrease computation time. This is because the coeff2LyA and coeff2X tables are all population-dependent. These individual operations also cost on the order of ~0.01-0.02 sec anyway, the most computationally taxing portion is the exp(gamma1 * gamma2 * xi) - 1 operation, which we can't take out of a loop or define external to the get_all_corrs functions.